### ISSUE TITLE
Unification errors when using lists in abilities

### ISSUE THREAD

**Issue by ChrisPenner** [unisonweb/unison#2298]
Okay… colour me confused on this one :upside_down_face:

`unison version: latest-137-g3b6880f2e'`

This doesn’t compile:
```
wat : '{Stream [x]} ()
wat _ =
  acc = [] -- unused, but relevant
  if true
    then
      ()
    else
      emit []
```

```ucm
  The expression in red needs the {Stream [elem17]} ability, but this location only has access to the {Stream [x6]} ability,

     31 |       emit []
```

If we remove the completely useless binding statement it compiles:

```
wat : '{Stream [x]} ()
wat _ =
  if true
    then
      ()
    else
      emit []
```

Or; if we add an ADDITIONAL emit outside of the if-block it works…

```
wat : '{Stream [x]} ()
wat _ =
  acc = []
  emit [] -- Emitting outside of the 'if' fixes the unification...?
  if true
    then
      ()
    else
      emit []
```

Not positive, but what I think is happening is that the [] literal is getting cached with a concrete type rather than a polymorphic one?

Which is also why this version works:

```
wat : '{Stream [x]} ()
wat _ =
  acc = []
  if true
    then
      ()
    else
      x : [x] — Removing the type annotation causes it to fail with the same error.
      x = []
      emit x
```

The fact that it’s mentioning `elem`  in the type error despite it not being mentioned anywhere suggests the internal polymorphic variable of `[]` is leaking out.

I was unable to reproduce if I removed the
* initial binding
* if statement
* or ability

**Comment by dolio** [comment]
Can you try this again with the latest version? The examples seem to be working for me now. I think the fix to #2297 may have also fixed this.

**Comment by pchiusano** [comment]
Confirmed that this is no longer present on latest trunk, going to close but feel free to reopen or file a new ticket if you spot other weirdness.
